src/sign/openssl.rs

```diff
@@ -1,58 +1,373 @@
 //! Key and Signer using OpenSSL.
+
 #![cfg(feature = "openssl")]
 #![cfg_attr(docsrs, doc(cfg(feature = "openssl")))]
 
+use core::fmt;
 use std::vec::Vec;
-use openssl::error::ErrorStack;
-use openssl::hash::MessageDigest;
-use openssl::pkey::{PKey, Private};
-use openssl::sha::sha256;
-use openssl::sign::Signer as OpenSslSigner;
-use unwrap::unwrap;
-use crate::base::iana::DigestAlg;
-use crate::base::name::ToDname;
-use crate::base::octets::Compose;
-use crate::rdata::{Ds, Dnskey};
-use super::key::SigningKey;
-
-
-pub struct Key {
-    dnskey: Dnskey<Vec<u8>>,
-    key: PKey<Private>,
-    digest: MessageDigest,
+
+use openssl::{
+    bn::BigNum,
+    pkey::{self, PKey, Private},
+};
+
+use crate::base::iana::SecAlg;
+
+use super::{generic, Sign};
+
+/// A key pair backed by OpenSSL.
+pub struct SecretKey {
+    /// The algorithm used by the key.
+    algorithm: SecAlg,
+
+    /// The private key.
+    pkey: PKey<Private>,
+}
+
+impl SecretKey {
+    /// Use a generic secret key with OpenSSL.
+    ///
+    /// # Panics
+    ///
+    /// Panics if OpenSSL fails or if memory could not be allocated.
```

Comment:
I don't think we want an entire name server to crash if there is a failure signing one zone so I'm not sure that this should panic if OpenSSL fails.

Reply:
OpenSSL exposes very little information about possible failures, but since we are only performing CPU-based operations here, the only reasonable failure case is out-of-memory -- which will probably cause panics anyway.  Issues with malformed keys should result in an `Err` instead, but that's hard to verify -- I might have to do something silly like sign a `[u8; 0]` here just to make OpenSSL perform those checks.

Reply:
The problem with panicking in a library is you don't know what context you are being used in. The signing might be a tiny unimportant part of an application with much more important objectives for which it is imperative that it not abort due to panicking. I'm also wary of assuming reasons why calls to OpenSSL may fail, are you sure you can say it is only OOM that will cause signing failure?

Reply:
I've removed the vast majority of panics and replaced them with opaque "implementation error" variants.